### PR TITLE
Interrupts

### PR DESCRIPTION
**Pull request by hamelin**

### CODE REVIEW

File: greensim/__init__.py [hunk 1]
```diff
@@ -71,6 +72,87 @@ def _log(self, level: int, event: str, **params: Any) -> None:
...
+            self.args == other.args and \
+            self.kwargs == other.kwargs
+
+    def __lt__(self, other: object) -> bool:
+        if not isinstance(other, _Event):
+            raise ValueError("Both terms of the comparison must be _Event instances.")
+        return (self._timestamp, self._identifier) < (other._timestamp, other._identifier)
+
+    @property
+    def timestamp(self) -> Optional[float]:
```
[the remark this answers is not in the document]
Comment (hamelin) [reply greensim/__init__.py]: Great suggestion. However, greensim runs on abstract time units: 1.0 time unit in greensim does not map to a second, or to an hour, or whatever. Thus, the mapping to real time is a task for the code that depends on greensim. I can think of one project that depends on greensim, and that's [ITsim](https://github.com/ElementAI/itsim/): this one enforces the convention that 1.0 matches 1 second. This project might then want to anchor the origin of its simulation to some date and time. Furthermore, this functionality might be mostly interesting in a post-simulation output analysis phase, so this time transformation may even be ad hoc to ITsim. So we can delay this concern to later.

File: greensim/__init__.py [hunk 2]
```diff
@@ -71,6 +72,87 @@ def _log(self, level: int, event: str, **params: Any) -> None:
...
+        self._timestamp = timestamp
+        self._identifier = identifier
+        self._is_cancelled = False
+        self._event = event
+        self._args = args
+        self._kwargs = kwargs
+
+    def __eq__(self, other: object) -> bool:
+        return isinstance(other, _Event) and \
+            self._timestamp == other._timestamp and \
```
[the remark this answers is not in the document]
Comment (hamelin) [reply greensim/__init__.py]: Fixed.

File: greensim/__init__.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (plbeaubien-eai) [reply greensim/__init__.py]: Makes sense! Thanks for the details

File: greensim/__init__.py [hunk 3]
```diff
@@ -71,6 +72,87 @@ def _log(self, level: int, event: str, **params: Any) -> None:
...
+    Raised on a :py:class:`Process` instance through the :py:meth:`Process.interrupt` method, so it resumes its
+    execution without having advanced in time as much as it expected, or having fulfilled the condition is hoped to
+    satisfy by going into pause.
+    """
+    def __eq__(self, other: object) -> bool:
+        return isinstance(other, Interrupt)
+
+
+@total_ordering
+class _Event(object):
```
[the remark this answers is not in the document]
Comment (hamelin) [reply greensim/__init__.py]: TIL. :smile: Fixed.

File: tests/test_logging.py [hunk 4]
```diff
@@ -207,12 +207,89 @@ def proc():
...
+
+    sim = Simulator()
+    proc_proc = sim.add(proc)
+    name_proc_orig = proc_proc.local.name
+    sim.run()
+
+    name_proc = proc_proc.local.name
+    check_log(
+        auto_logger,
+        (logging.INFO, 0.0, "", "Simulator", sim.name, "add", dict(fn=proc, args=(), kwargs={})),
```
[the remark this answers is not in the document]
Comment (hamelin) [reply tests/test_logging.py]: Simulation do have a UUID -- it's their `name` attribute. Every subclass of `Named` (which is more a trait than a subclass) carries as default name a random UUID, unless they are given an explicit name on instantiation.

**Comment by hamelin** [comment]
@plbeaubien-eai Can you take a look at the fixes according to your suggestions?

File: greensim/__init__.py [hunk 5]
```diff
@@ -71,6 +72,87 @@ def _log(self, level: int, event: str, **params: Any) -> None:
...
+            self.timestamp == other.timestamp and \
+            self.identifier == other.identifier and \
+            self.fn == other.fn and \
+            self.args == other.args and \
+            self.kwargs == other.kwargs
+
+    def __lt__(self, other: object) -> bool:
+        if not isinstance(other, _Event):
+            raise ValueError("Both terms of the comparison must be _Event instances.")
+        return (self._timestamp, self._identifier) < (other._timestamp, other._identifier)
```
Comment (plbeaubien-eai) [inline greensim/__init__.py]: Same as above... other._timestamp to other.timestamp

Comment (hamelin) [reply greensim/__init__.py]: Fixed. Please check again @plbeaubien-eai

Comment (hamelin) [reply greensim/__init__.py]: Doh! not fixed yet. Give me a sec.

Comment (hamelin) [reply greensim/__init__.py]: Ok, ready now.

**Pull request merged**